File: go/options/zadd_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import (
	"errors"

	"github.com/valkey-io/valkey-glide/go/v2/constants"

	"github.com/valkey-io/valkey-glide/go/v2/internal/utils"
)

// Optional arguments to `ZAdd` in [SortedSetCommands]
type ZAddOptions struct {
	ConditionalChange constants.ConditionalSet
	UpdateOptions     UpdateOptions
	Changed           bool
	Incr              bool
	Increment         float64
	Member            string
}

func NewZAddOptions() *ZAddOptions {
	return &ZAddOptions{}
}

// `conditionalChange` defines conditions for updating or adding elements with `ZADD` command.
func (options *ZAddOptions) SetConditionalChange(c constants.ConditionalSet) *ZAddOptions {
	options.ConditionalChange = c
	return options
}

// `updateOptions` specifies conditions for updating scores with zadd command.
func (options *ZAddOptions) SetUpdateOptions(u UpdateOptions) *ZAddOptions {
	options.UpdateOptions = u
	return options
}

// `Changed` changes the return value from the number of new elements added to the total number of elements changed.
func (options *ZAddOptions) SetChanged(ch bool) (*ZAddOptions, error) {
	if options.Incr {
		return nil, errors.New("changed cannot be set when incr is true")
	}
	options.Changed = ch
	return options, nil
}

// `INCR` sets the increment value to use when incr is true.
func (options *ZAddOptions) SetIncr(incr bool, increment float64, member string) (*ZAddOptions, error) {
	if options.Changed {
		return nil, errors.New("incr cannot be set when changed is true")
	}
	options.Incr = incr
	options.Increment = increment
	options.Member = member
	return options, nil
}

// `ToArgs` converts the options to a list of arguments.
func (opts *ZAddOptions) ToArgs() ([]string, error) {
	args := []string{}
	var err error

	if opts.ConditionalChange == constants.OnlyIfExists || opts.ConditionalChange == constants.OnlyIfDoesNotExist {
		args = append(args, string(opts.ConditionalChange))
	}

	if opts.UpdateOptions == ScoreGreaterThanCurrent || opts.UpdateOptions == ScoreLessThanCurrent {
		args = append(args, string(opts.UpdateOptions))
	}

	if opts.Changed {
		args = append(args, constants.ChangedKeyword)
	}

	if opts.Incr {
		args = append(args, constants.IncrKeyword, utils.FloatToString(opts.Increment), opts.Member)
	}

	return args, err
}

type UpdateOptions string

const (
	// Only update existing elements if the new score is less than the current score. Equivalent to
	// "LT" in the Valkey API.
	ScoreLessThanCurrent UpdateOptions = "LT"
	// Only update existing elements if the new score is greater than the current score. Equivalent
	// to "GT" in the Valkey API.
	ScoreGreaterThanCurrent UpdateOptions = "GT"
)

File: csharp/sources/Valkey.Glide/Commands/Options/ZAddOptions.cs
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

namespace Valkey.Glide.Commands.Options;

/// <summary>
/// Defines conditions for updating or adding elements with the ZAdd command.
/// </summary>
public enum ConditionalSet
{
    /// <summary>
    /// Only update existing elements. Equivalent to "XX" in the Valkey API.
    /// </summary>
    OnlyIfExists,
    
    /// <summary>
    /// Only add new elements. Equivalent to "NX" in the Valkey API.
    /// </summary>
    OnlyIfDoesNotExist
}

/// <summary>
/// Specifies conditions for updating scores with the ZAdd command.
/// </summary>
public enum UpdateOptions
{
    /// <summary>
    /// Only update existing elements if the new score is less than the current score. Equivalent to "LT" in the Valkey API.
    /// </summary>
    ScoreLessThanCurrent,
    
    /// <summary>
    /// Only update existing elements if the new score is greater than the current score. Equivalent to "GT" in the Valkey API.
    /// </summary>
    ScoreGreaterThanCurrent
}

/// <summary>
/// Optional arguments for the ZAdd command.
/// </summary>
public class ZAddOptions
{
    /// <summary>
    /// Defines conditions for updating or adding elements.
    /// </summary>
    public ConditionalSet? ConditionalChange { get; private set; }

    /// <summary>
    /// Specifies conditions for updating scores.
    /// </summary>
    public UpdateOptions? UpdateOption { get; private set; }

    /// <summary>
    /// Changes the return value from the number of new elements added to the total number of elements changed.
    /// </summary>
    public bool Changed { get; private set; }

    /// <summary>
    /// Sets the conditional change option.
    /// </summary>
    /// <param name="conditionalChange">The conditional change option.</param>
    /// <returns>This ZAddOptions instance for method chaining.</returns>
    public ZAddOptions SetConditionalChange(ConditionalSet conditionalChange)
    {
        ConditionalChange = conditionalChange;
        return this;
    }

    /// <summary>
    /// Sets the update option.
    /// </summary>
    /// <param name="updateOption">The update option.</param>
    /// <returns>This ZAddOptions instance for method chaining.</returns>
    public ZAddOptions SetUpdateOption(UpdateOptions updateOption)
    {
        UpdateOption = updateOption;
        return this;
    }

    /// <summary>
    /// Sets the changed flag.
    /// </summary>
    /// <param name="changed">Whether to return the number of changed elements instead of added elements.</param>
    /// <returns>This ZAddOptions instance for method chaining.</returns>
    public ZAddOptions SetChanged(bool changed)
    {
        Changed = changed;
        return this;
    }

    /// <summary>
    /// Converts the options to a list of arguments for the ZAdd command.
    /// </summary>
    /// <returns>A list of string arguments.</returns>
    internal List<GlideString> ToArgs()
    {
        var args = new List<GlideString>();

        if (ConditionalChange.HasValue)
        {
            args.Add(ConditionalChange.Value switch
            {
                ConditionalSet.OnlyIfExists => "XX",
                ConditionalSet.OnlyIfDoesNotExist => "NX",
                _ => throw new ArgumentException($"Unknown ConditionalSet value: {ConditionalChange.Value}")
            });
        }

        if (UpdateOption.HasValue)
        {
            args.Add(UpdateOption.Value switch
            {
                UpdateOptions.ScoreLessThanCurrent => "LT",
                UpdateOptions.ScoreGreaterThanCurrent => "GT",
                _ => throw new ArgumentException($"Unknown UpdateOptions value: {UpdateOption.Value}")
            });
        }

        if (Changed)
        {
            args.Add("CH");
        }

        return args;
    }
}
